epic/new hooks

kmaschi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

This PR includes features and fixes for the plugin-hooks repository.
- Migration of project from JS to TS.
  - Includes developer tooling, actions, linting, and test coverage.
  - Builds CJS and ESM for NPM use cases.
- Adds afterAll hook
- Adds beforeSource/afterSource hooks 

## Description

<!--- Describe your changes in detail -->

Current state of this project is in JS with very minimal/non-existent test coverage. Moving forward we would like to have confidence in changes through test coverage. The move to TypeScript will improve developer experience working within the package and provide useful typings that can be used by users creating hooks.

BeforeSource hook executes once before the fetch to a named source. It has access to the request allowing users to perform actions such as modifying request headers.

AfterSource hook executes once after the fetch to a named source. It has access to the response allowing users to perform actions such as modifying response headers.

AfterAll hook executes once before the final response is sent to the client.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

CEXT-1441

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
Goals of better developer and user experience.
- Better developer experience in TypeScript.
- Test coverage
- Prettier/linting for code quality/consistency.
- Typings for developers.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Changes lint, test, and build successful in GitHub Actions
- Published an alpha to NPM [0.3.5-alpha.4](https://www.npmjs.com/package/@adobe/plugin-hooks/v/0.3.5-alpha.4)

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
chore: - Migrated js to typescript
- Added package build configuration for esm/cjs

Commit:
chore: - Updated to support better build types.
- Added tests.
- Refactoring for tests.

Commit:
chore: - Updated to rimraf@5.0.1 for node18 compatibility

Commit:
chore: - Updated to alpha package version.

kmaschi committed:
Merge pull request #31 from adobe/chore/ts-port

Chore/ts port

Commit:
chore: - Updated tests to ensure expected payload passed into hook functions.

kmaschi committed:
Merge pull request #32 from adobe/chore/ts-port

chore: - Updated tests to ensure expected payload passed into hook fu…

Commit:
chore: - Updated readme.

Commit:
chore: - Updated local/remote usage in readme.

Commit:
chore: - Updated index to export useful types for development.

Commit:
chore: - Enhanced README with detailed usage instructions and notes on hooks

kmaschi commented:
Created as epic to test NPM publish.

Commit:
chore: - Moved export types.

dani97 commented on src/utils.ts:
```diff
@@ -0,0 +1,321 @@
+ * @param modulePath Module path.
+ */
+export async function importFn(modulePath: string) {
+	return Promise.resolve(import(modulePath)).then(module => module);
```
@kmaschi @nvyasadobe This function is translated to dynamic require in cjs. If the source code calling this function is esm you will not have any problem but if cjs code calls the dynamic require (cjs translation). For now I am updating the cjs folder every time I compile. 

We can separate this function to a new file and it should be in js so won't be compiled from ts to js

dani97 committed:
fix: cext-1411: hooks cjs fix

dani97 committed:
fix: cext-1441: dyn import

dani97 committed:
chore: cext-0000: lint error

dani97 committed:
chore: cext-0000: fix typo

nvyasadobe committed:
feat: first commit for afterAll implementation

dani97 committed:
chore: cext-0000: fix tests

nvyasadobe committed:
feat: WIP: after hook implementation local hook and remote hook

nvyasadobe committed:
feat: added test coverage for afterAll

nvyasadobe committed:
chore: bump alpha version

nvyasadobe committed:
chore: exported error type from graphQl

dani97 committed:
feat: cext-1441: after before source hooks

nvyasadobe committed:
chore: address some of review comments

nvyasadobe committed:
chore: address review comments: moved logic of afterAll and beforeAll outside of index.ts

nvyasadobe committed:
chore: address review comments removed error codes seperately

nvyasadobe committed:
chore: updated tests

nvyasadobe committed:
chore: updated package.json to relase alpha.1

nvyasadobe committed:
chore: minor correction

nvyasadobe committed:
chore: did some refractor and updated error logic moved to common utils

nvyasadobe committed:
chore: created hookResolver to handle commanality between afterAll and beforeAll

dani97 committed:
chore: cext-1441: code improvements linters types fix

dani97 committed:
chore: cext-1441: types fix

nvyasadobe committed:
Merge pull request #34 from adobe/afterall-hooks-implementation

Afterall hooks implementation

dani97 committed:
tests: cext-1441: add tests

dani97 committed:
Merge branch 'epic/ts-migration' of github.com:adobe/plugin-hooks into cext-1441-after-before-source-hooks

dani97 committed:
tests: cext-1441: fix tests

nvyasadobe committed:
chore: added dynamicImport.js to handle the commonJs case as well and failing build on node version 18

nvyasadobe committed:
Merge pull request #37 from adobe/fix-syntaxError-dynamicimport

chore: added dynamicImport.js to handle the commonJs case (failing on node v.18 inside docker)

dani97 committed:
fix: cext-1441: update After Response playload interface

dani97 committed:
fix: cext-1441: update version

dani97 committed:
Merge branch 'epic/ts-migration' of github.com:adobe/plugin-hooks into cext-1441-after-before-source-hooks

dani97 committed:
fix: cext-1441: hooks

dani97 committed:
chore: cext-1441: update version

dani97 committed:
Merge pull request #35 from adobe/cext-1441-after-before-source-hooks

Cext 1441 after before source hooks

dani97 committed:
fix: cext-1441: fix source confusion issue

dani97 committed:
fix: cext-1441: update package

dani97 committed:
Merge pull request #38 from adobe/cext-1441-after-before-source-hooks

fix: cext-1441: fix source hooks mis config issue

dani97 committed:
chore: cext-1441: release stable version

Commit:
feat(context-state): CEXT-4829 - Extracted hook function import logic for reuse.
- Added secrets, state, and logger to context in all hooks(beforeAll, afterAll, beforeSource, afterSource).
- Resolved linting/style warnings.

Commit:
feat(context-state): CEXT-4829 - Extracted hook function import logic for reuse.
- Added secrets, state, and logger to context in all hooks(beforeAll, afterAll, beforeSource, afterSource).
- Resolved linting/style warnings.

Commit:
feat(context-state): CEXT-4829 - Added extensions to beforeSource/afterSource errors.

Commit:
feat(context-state): CEXT-4829 - Fixed linting errors.

kmaschi committed:
Merge pull request #39 from adobe/cext-1441-after-before-source-hooks

Cext 1441 after before source hooks

Commit:
feat(context-state): CEXT-4829 - Updated package version.

kmaschi committed:
Merge branch 'epic/ts-migration' into feature/CEXT-4829-context-state

amolina-adobe commented on package.json:
```diff
@@ -1,32 +1,71 @@
     "graphql": "^16.6.0",
-    "jest": "^29.6.4"
+    "graphql-yoga": "5.3.1",
+    "jest": "^29.6.4",
```
Still need jest?

amolina-adobe commented on src/handleAfterSourceHooks.ts:
```diff
@@ -0,0 +1,71 @@
+
+			if (hookFn) {
+				try {
+					const hooksResponse = await hookFn(payload);
```
More of a side note, but I don't really understand the payload passed in here. Looks like we're supplying a AfterSourceHookPayload, yet the function expects a SourceHookFunctionPayload. Looking at the types, I believe the AfterSourceHookPayload still satisfies the SourceHookFunctionPayload, but it's not immediately apparent what the distinction is.

amolina-adobe commented on src/handleBeforeSourceHooks.ts:
```diff
@@ -0,0 +1,71 @@
+						}
+					}
+				} catch (err: unknown) {
+					logger.error('Error while invoking beforeSource hook %o', err);
+					if (err instanceof Error) {
+						throw new Error(err.message);
+					}
+					if (err && typeof err === 'object' && 'message' in err) {
+						throw new Error((err as { message?: string }).message);
+					}
+					throw new Error('Error while invoking beforeSource hook');
+				}
```
Minor, but I'm thinking this can re-use some of the logic in errorHandler.ts

amolina-adobe commented on src/handleAfterSourceHooks.ts:
```diff
@@ -0,0 +1,71 @@
+import type { YogaLogger } from 'graphql-yoga';
```
Do we need to add the copyright to all files?

amolina-adobe commented on src/handleBeforeSourceHooks.ts:
```diff
@@ -0,0 +1,71 @@
+import type { YogaLogger } from 'graphql-yoga';
```
Copyright?

Commit:
chore(plugin-hooks-fixes): - Updated payloads to have more consistent types.
- Updated remote hook payload to only include appropriate serializable context.

Commit:
chore(plugin-hooks-fixes): - Updated error handling in source hooks.
- Moved error handling related functions to error module.
- Added copyright notices in new files.

Commit:
chore(plugin-hooks-v2): - Removed unnecessary functions.
- Added jsdocs.

Commit:
chore(plugin-hooks-fixes): - Fixed tests.

Commit:
chore(plugin-hooks-fixes): - Removed jest.

Commit:
chore(plugin-hooks-fixes): - Updated to new alpha version

kmaschi committed:
Merge pull request #40 from adobe/feature/CEXT-4829-context-state

Feature/cext 4829 context state

kmaschi committed:
Merge pull request #41 from adobe/chore/plugin-hooks-fixes

Chore/plugin hooks fixes

nvyasadobe committed:
chore: WIP: afterSource hook fix for remote hook

nvyasadobe committed:
feat: fixed the beforeSource remote hooks

nvyasadobe committed:
feat: minor correction and updated headers merge logic

nvyasadobe committed:
feat: added extensions support for afterAll

Commit:
chore(remote-fix) - Removed unnecessary executors since invocated differently.
- Updated types to make clear the properties available to each hook handler.
- Handled stacked beforeAll/afterAll hooks ensuring both local and remove functions can work with serialized payload.
- Handled stacked beforeAll/afterAll hooks ensuring handlers can work with serialized response data.
- Updated tests.

kmaschi committed:
Merge pull request #42 from adobe/chore/sourcehooks-remote-fix

Chore/sourcehooks remote fix

Commit:
feat(plugin-hooks): - Bumped minor version.

brasewel committed:
Updating readme. Excluding examples

brasewel committed:
Merge pull request #43 from adobe/chore/update-readme

Updating readme. Excluding examples
